### PR TITLE
allow number of split particles to be configurable

### PR DESCRIPTION
**Pull request by Dekken**
suggestive changes, alternatives welcome

https://github.com/PHAREHUB/PHARE/issues/162

### CODE REVIEW

**Comment by nicolasaunai** [comment]
- so here I think we need to get the number from the dict which takes it from pharein
- second, it would be nice if the `nbRefinedPart` was not a Gridlayout template attribute, because a priori GridLayout deals with meshes and has nothing to do with splitting parameters. 

I'm thinking out loud : 

ideally the `nbRefinedPart` should be a Split template param and Split could forward the constexpr value so that : 

```
ParticlesRefineOperator<dimension, interpOrder, ParticlesDataSplitType::interior,
                                      nbRefinedPart, SplitT>;
```

can just write : 

```
ParticlesRefineOperator<dimension, interpOrder, ParticlesDataSplitType::interior,
                                      nbRefinedPart>;
```


I know the hybhybstrat still has this

```
using SplitT                               = Split<dimension, interpOrder>;
```

and so that would become 
```
using SplitT                               = Split<dimension, interpOrder, nbRefinePart>;
```
which would mean we still have to get nbRefinePart from somewhere...
in fact I think it's not even the HybHyb strat business to know what `Split` is... that should be internal detail to the refinement operators.

Maybe the HybridHybridMessengerStrategy should have one more template param that would be something like a RefinementParams that would offer : 

```
InteriorParticleRefineOp
CoarseToFineRefineOpOld
CoarseToFineRefineOpNew
```

types.

The MessengerFactory would create that struct and pass it as a template param of the hybhyb.
The messenger factory itself is defined in PHARETypes where it could be defined now with the nbRefinedPart.


something like that...

**Comment by Dekken** [comment]
I don't see how we can propagate the runtime python value to the compile template

**Comment by nicolasaunai** [comment]
we do that already for dim and inter_order though...

**Comment by Dekken** [comment]
> we do that already for dim and interp_order though...

like what I've done perhaps?

**Comment by nicolasaunai** [comment]
well like what I did for make simulator and that you did for the diagnostics, the Maker trick, no?

**Comment by Dekken** [comment]
> well like what I did for make simulator and that you did for the diagnostics, the Maker trick, no?

@Dekken Dekken force-pushed the Dekken:splitNbr branch from 279fbb4 to 88267ac 3 hours ago

**Comment by Dekken** [comment]
changed how makeAtRuntime was working as I was having issues with resolving different lists of possible number of split particles per dim/interp

the client code becomes a bit simpler as a result I think tho 
https://github.com/Dekken/PHARE/blob/splitNbr/src/python3/data_wrangler.cpp#L455

**Comment by Dekken** [comment]
@nicolasaunai feel free to have a look and tell me what you think

File: tests/simulator/py/__init__.py [hunk 1]
```diff
@@ -23,6 +23,7 @@ def basicSimulatorArgs(dim:int, interp:int, **kwargs):
         "dl": dl,
         "max_nbr_levels": 2,
         "refinement_boxes": {"L0": {"B0": [(10,), (50,)]}},
+        "refined_particle_nbr": dim * 2,
```
[the remark this answers is not in the document]
Comment (Dekken) [reply tests/simulator/py/__init__.py:26]: this will have to be reviewed for 2/3d anyway

**Pull request merged**